Remove php_error_docref, AZURE_LOG_SYS and update document (#23)
microsoft/mysqlnd_azure: doc/mysqlnd_azure_log.md

# Diagnostic Log Configuration

Log down some runtime information from mysqlnd\_azure.

## Configuration

> 

Reviewer: Add a single line here on what this section is about

Author: Added. I also refactor the explanation details section for better perception, and for test scenario check. Can you help have another look too?

> ### Log destination supported (mysqlnd\_azure.logOutput)

Reviewer: To keep the headings on this page consistent for all the configuration. Change this to

 ###mysqlnd\_azure.logOutput
Use this property to set the destination path for your logs. For all the file paths mentioned below, except Apache and php-fpm's system error log, it need to be configured with right user permission, e.g. www-data for Apache. For relative file path, when working in cli, the father path is the current working directoy, when working with Apache, the father path is /.   We provide multiple options to store your logs .

**Notice:**

> For all the file paths mentioned below, except Apache and php-fpm's system error log, it need to be configured with right user permission, e.g. www-data for Apache. For relative file path, when working in cli, the father path is the current working directoy, when working with Apache, the father path is /.

Reviewer: By including this section as per my previous comment , you dont need this as a notice

We provide multiple options for log destination you may want to choose. 

> Name |mysqlnd\_azure.logOutput

Reviewer: This table format is confusing  , can we simplify this as 

 mysqlnd_azure.logOutput= 0 

0 is the default value but it can accept the values 1,2 and 4.  See details below for each value and choose the one based on your needs.  

Now show the Value/Behavior Table at line 23

:----- | :------

> Comment | Destination for log

Reviewer: Change "comment" to "Description" . For all the properties below

Type | Enumeration
Accepted Value | [ 0\| 1\| 2 \| 4 ]
Default | 0 (No log)
Dynamic | No

Value | Behavior | Sample log
:----- | :------ | :----- 
0 | Like logLevel control, it will turn off the log. No log will be print out. | 

> 1 | Print logs using PHP error log API. <br/> - PHP CLI: Print to console. <br/> - Work scenario with Apache/Nginx: Will be discussed in seperated section below. |  Warning: mysqli_real_connect(): [2020-07-20 14:09:03] [MYSQLND_AZURE] [ERROR] CLIENT_SSL is not set when mysqlnd_azure.enableRedirect is ON in mysqli_basic.php on line 6

Reviewer 2: ```suggestion
1 | Print logs using PHP error log API. <br/> - PHP CLI: Print to console. <br/> - Works with Apache/Nginx: Will be discussed in separate section below. |  Warning: mysqli_real_connect(): [2020-07-20 14:09:03] [MYSQLND_AZURE] [ERROR] CLIENT_SSL is not set when mysqlnd_azure.enableRedirect is ON in mysqli_basic.php on line 6
```

2 | Print logs to a pre-defined file by mysqlnd\_azure's config mysqlnd\_azure.logfilePath=xx.log. | [MYSQLND_AZURE] [ERROR] CLIENT_SSL is not set when mysqlnd_azure.enableRedirect is ON

> 4 | Print logs using stderr. <br/> - PHP CLI: Print to console. <br/> - Work scenario with Apache/Nginx: Will be discussed in seperated section below. | [MYSQLND_AZURE] [ERROR] CLIENT_SSL is not set when mysqlnd_azure.enableRedirect is ON

Reviewer 2: ```suggestion
4 | Print logs using stderr. <br/> - PHP CLI: Print to console. <br/> - Works with Apache/Nginx: Will be discussed in separate section below. | [MYSQLND_AZURE] [ERROR] CLIENT_SSL is not set when mysqlnd_azure.enableRedirect is ON
```

Reviewer: +1 to Andrea comments

Author: I have refactor the structure for better perception and test check, can you help have another look?

> #### Detailed behavior for logOutput1 and logOutput4

Reviewer 2: ```suggestion
#### Detailed behavior for logOutput = 1 and logOutput=4
```

Reviewer: +1 to Andrea comments

> - **For mod-php and Apache (take php7.2 as example)**

Reviewer: change to **For mod-php and Apache (using PHP 7.2 as an example )**

Config file | Config options | Usage
:----- | :------ | :----- 

> /etc/php/7.2/apache2/php.ini | error_log <br/> log_error <br/> display_errors   | This file is Apache mod_php's config file s <br/> -log_error: logOutput1 is only available if this is on. <br/> error_log: where logOutput1 goes into. <br/> --1.	error_log=syslog:  logOutput1 is not available with this option. <br/> --2.	error_log=A_file_path. It will take priority of apache’s error log setting. <br/> display_errors: Whether print logOutput1 to a client web page.

Reviewer 2: ```suggestion
/etc/php/7.2/apache2/php.ini | error_log <br/> log_error <br/> display_errors   | This file is Apache mod_php's config file<br/> -log_error: logOutput=1 is only available if this is on. <br/> error_log: where logOutput=1 goes into. <br/> (1)	error_log=syslog:  logOutput= 1 is not available with this option. <br/>(2) error_log=A_file_path. It will take priority of Apache’s error log setting. <br/> display_errors: Whether print logOutput=1 to a client web page.
```

> /etc/apache2/sites-available/000-default.conf | ErrorLog | This file is Apache's error log path for seperated virtual instance. <br/> If error_log is not set in php.ini above, logOutput1 will go here.

Reviewer 2: ```suggestion
/etc/apache2/sites-available/000-default.conf | ErrorLog | This file is Apache's error log path for separate virtual instance. <br/> If error_log is not set in php.ini above, logOutput=1 will output here.
```

> /etc/apache2/apache2.conf | ErrorLog | This file is Apache’s global error log path. <br/> logOutput4 goes here. <br/> If ErrorLog is not set for each virtual host, and error_log is not set in php.ini, then logOutput1 also goes here.

Reviewer 2: ```suggestion
/etc/apache2/apache2.conf | ErrorLog | This file is Apache’s global error log path. <br/> logOutput=4 outputs here. <br/> If ErrorLog is not set for each virtual host and error_log is not set in php.ini, then logOutput=1 also outputs here.
```

>  - **For php-fpm and Apache (take php7.2 as example)**

Reviewer 2: ```suggestion
 - **For PHP-FPM and Apache (take PHP7.2 as example)**
```

Reviewer: Change this "take PHP7.2 as example" to "using PHP 7.2 as example "

>  

Reviewer 2: Nginx with PHP7.2 also works similarly as the below behavior.

Author: I have update with follows: "Note: Nginx with PHP-FPM works similarly as the below corrensponding behavior." before the section. Please check.

 Config file | Config options | Usage
:----- | :------ | :----- 

> /etc/php/7.2/fpm/php-fpm.conf | error_log	| This Php-fpm’s system error log, has content like php-fpm is restarted or crashed. <br/> -- logOutput4 will go here if catch_workers_output = yes in www.conf. <br/> logOutput1 will also go here if catch_workers_output = yes in www.conf && log_errors=ON && error_log is not set in php.ini and www.conf below.

Reviewer 2: ```suggestion
/etc/php/7.2/fpm/php-fpm.conf | error_log	| This is PHP-FPM's system error log and contains content such as PHP-FPM's state (ex. restarted or crashed). <br/> - logOutput=4 will output logs here if catch_workers_output = yes in www.conf. <br/> logOutput=1 will also log here if catch_workers_output = yes in www.conf && log_errors=ON && error_log is not set in php.ini and www.conf below.
```

> /etc/php/7.2/fpm/php.ini | error_log <br/> log_error <br/> display_errors	| - error_log: The destination logOutput1 goes to. <br/> - log_error: same with mod_php. <br/> - display_errors: same with mod_php.

Reviewer 2: ```suggestion
/etc/php/7.2/fpm/php.ini | error_log <br/> log_error <br/> display_errors	| - error_log: The destination logOutput=1 logs to. <br/> - log_error: same with mod_php. <br/> - display_errors: same with mod_php.
```

> /etc/php/7.2/fpm/pool.d/www.conf	| catch_workers_output <br/> php_admin_value[error_log] <br/> php_admin_flag[log_errors] <br/> php_flag[display_errors] | - catch_workers_output: <br/> Whether catch logOutput4 and logOutput1. <br/> <br/> -- php_admin_value[error_log]: <br/> -- php_admin_flag[log_errors]: <br> -- php_flag[display_errors]: <br/> Same as that in php.ini, but the definitions here take priority.

Reviewer 2: ```suggestion
/etc/php/7.2/fpm/pool.d/www.conf	| catch_workers_output <br/> php_admin_value[error_log] <br/> php_admin_flag[log_errors] <br/> php_flag[display_errors] | - catch_workers_output: <br/> Whether to catch logOutput=4 and logOutput=1. <br/> <br/> -- php_admin_value[error_log]: <br/> -- php_admin_flag[log_errors]: <br> -- php_flag[display_errors]: <br/> Same as that in php.ini, but the definitions here take priority.
```

> /etc/apache2/sites-available/000-default.conf	| ErrorLog	| logOutput1 will go here if the error_log is not set in php.ini and www.conf. <br/> Please notice that the logs for single request is in single line since fastcgi treat one request’s log as single line)

Reviewer 2: ```suggestion
/etc/apache2/sites-available/000-default.conf	| ErrorLog	| logOutput=1 will output here if the error_log is not set in php.ini and www.conf. <br/> Please notice that the logs for single request is in single line since FastCGI treats one request’s log as single line)
```

> /etc/apache2/apache2.conf	| ErrorLog	| Apache’s global error log path. <br/> logOutput1 will go here if the error_log is not set in php.ini and www.conf and ErrorLog is not set in each virtual instance

Reviewer 2: ```suggestion
/etc/apache2/apache2.conf	| ErrorLog	| Apache’s global error log path. <br/> logOutput=1 will go here if the error_log is not set in php.ini and www.conf and ErrorLog is not set in each virtual instance
```

> Work scenario with Nginx is similar.

Reviewer 2: Probably want to move this line close to the title of the table. 

```suggestion
```

 ### mysqlnd\_azure.logfilePath

Name | mysqlnd\_azure.logfilePath
:----- | :------

> Comment | Filename that you want your log writes to, if mysqlnd\_azure.logOutput=2

Reviewer 2: ```suggestion
Comment | Filename that you want your log to write to if mysqlnd\_azure.logOutput=2
```

Reviewer: Change Comment to Description here . It is confusing with the word Comment.

Type | String

> Accepted Value | String length <= 255, legal filename string

Reviewer 2: ```suggestion
Accepted Value | String length <= 255, legal filename string.<br>Length cannot exceed the file system's restriction (255). Otherwise, system will use the default file name and throw a warning "Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log
```

Default | mysqlnd\_azure\_runtime.log
Dynamic | NO

> Note | Filename length cannot exceed the file system's restriction(255), otherwise system will use the default filename and throw a warning "Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log"

Reviewer 2: Move this line to the first "Accepted value" line.

```suggestion
```

 ### mysqlnd\_azure.logLevel
Name | mysqlnd\_azure.logLevel
:----- | :------

> Comment | The verbose level of log we generate.

Reviewer: change comment to description

Type | Enumeration
Accepted Value | [ 0 \| 1 \| 2 \| 3 ]
Default | 0 (OFF)
Dynamic | Yes

> Note | This variable is dynamic, you may use user srcipt API `ini_set('mysqlnd_azure.logLevel', '$VAL'); ` to check log with different verbose level.

Reviewer 2: Move this line to the same 
```suggestion
Note | This variable is dynamic, you may use user script API `ini_set('mysqlnd_azure.logLevel', '$VAL'); ` to check log with different verbose level.
```

### Loglevel supported

> we support 3 level of logs for user:

Reviewer 2: ```suggestion
The mysqlnd_azure extension supports three log levels:
```

- [DEBUG]:

>   1. source code path(inside which function);

Reviewer 2: ```suggestion
  1. source code path (inside which function);
```

>   2. connection information(except password);

Reviewer 2: ```suggestion 
  2. connection information (except password);
```

  3. verbose running status (e.g. customer given redirected server information directly... )
- [INFO]:
  1. connection fallback when mysqlnd\_azure.enableRedirection=preferred with details (cannot get redirection
      info from OK packet/ try connect using redirect information failed , ...);
  2. cache operations
- [ERROR]:
  1. any operation may cause customer see connection break (e.x. original conn/ redirected\_conn
    established failed, ... and connection fall back when mysqlnd\_azure.enableRedirection = ON)

> > [SYSTM] maked log rows for Azure mysqlnd Log Module itself, e.g. invalid file length:

Reviewer 2: Assuming this was meant to be part of bullet list? 

```suggestion
- [SYSTM]: log rows for mysqlnd_azure log module itself, e.g. invalid file length:
```

> ```
> 2020-03-27 09:09:05 [SYSTM] Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log
> ```

> > note: OOM not included this version.

Reviewer 2: ```suggestion
> note: out of memory errors are not included this version of mysqlnd_azure.
```

#### Level value relationship to level logged
- 0: No log written
- 1: [SYSTM] + [ERROR]
- 2: [SYSTM] + [ERROR] + [INFO]
- 3: [SYSTM] + [ERROR] + [INFO] + [DEBUG]

## Usage Example

> > You can add to section [mysqlnd\_azure] in file `php.ini`

Reviewer 2: ```suggestion
> You can add to section [mysqlnd\_azure] in file `php.ini`. In the below example, the logs are output to a pre-defined file (using logOutput = 2) called "test.log" (using logfilePath). The logs level output includes system, error, info, and debug logs (set using logLevel = 3).

the logs are output to a pre-defined file called "test.log".  
```

Author: I have updated the test setting a bit to use absolute file path which is more recommended. Please help check.

:----- | :------ | :-----
A file path specifed in config files with different priorities. | Following files and options control where the log logs to, **prioity is from highest to lowerest**: <br/><br/> (1) **error_log** in /etc/php/7.2/apache2/php.ini <br/><br/> (2) **ErrorLog** in /etc/apache2/sites-available/000-default.conf (This file is Apache's error log path for seperated virtual instance if you define it). <br/><br/> (3) **ErrorLog** in /etc/apache2/apache2.conf (This file is Apache’s global error log path. If ErrorLog is not set for each virtual host, and error_log is not set in php.ini, then log goes here). | **log_errors=on** in /etc/php/7.2/apache2/php.ini
Client web page | **display_errors=on** in /etc/php/7.2/apache2/php.ini |  (1) **log_errors=on** in /etc/php/7.2/apache2/php.ini <br/><br/> (2) **display_errors=on** in /etc/php/7.2/apache2/php.ini. <br/><br/> Please notice this destination is not exclusive to a file path, you may have both.

Note: Nginx with PHP-FPM works similarly as the below corrensponding behavior.
 
 - **For PHP-FPM and Apache (using PHP 7.2 as an example)**
 
 Destination | Config file and options | Turn on switch
:----- | :------ | :----- 

> A file path specifed in config files with different priorities | Following files and options control where the log logs to, **prioity is from highest to lowerest**: <br/> (1) **php_admin_value[error_log]** in /etc/php/7.2/fpm/pool.d/www.conf <br/><br/> (2) **error_log** in /etc/php/7.2/fpm/php.ini <br/><br/> (3) **ErrorLog** in /etc/apache2/sites-available/000-default.conf <br/>Please notice that the logs for single request is in single line since FastCGI treat one request’s log as single line. <br/><br/> (4) **ErrorLog** in /etc/apache2/apache2.conf | (1) **catch_workers_output = yes** in  /etc/php/7.2/fpm/pool.d/www.conf <br/><br/> (2) Following files and options control the another switch, prioity is from highest to lowerest:<br/> (2-1) **php_admin_flag[log_errors]** in /etc/php/7.2/fpm/pool.d/www.conf <br/> (2-2) **log_errors** in /etc/php/7.2/fpm/php.ini

Reviewer 2: ```suggestion
A file path specifed in config files with different priorities | Following files and options control where the log outputs to, **priority is from highest to lowest**: <br/> (1) **php_admin_value[error_log]** in /etc/php/7.2/fpm/pool.d/www.conf <br/><br/> (2) **error_log** in /etc/php/7.2/fpm/php.ini <br/><br/> (3) **ErrorLog** in /etc/apache2/sites-available/000-default.conf <br/>Please notice that the logs for single request is in single line since FastCGI treats one request’s log as single line. <br/><br/> (4) **ErrorLog** in /etc/apache2/apache2.conf | (1) **catch_workers_output = yes** in  /etc/php/7.2/fpm/pool.d/www.conf <br/><br/> (2) Following files and options control the another switch, priority is from highest to lowest:<br/> (2-1) **php_admin_flag[log_errors]** in /etc/php/7.2/fpm/pool.d/www.conf <br/> (2-2) **log_errors** in /etc/php/7.2/fpm/php.ini
```

Author: Hi Andrea, 
The file you checked may not get update. The part for option using PHP API php_error_docref is completely removed now. Please check the latest version.

> Client web page | **display_errors=on** in /etc/php/7.2/apache2/php.ini or in /etc/php/7.2/fpm/pool.d/www.conf | (1) **log_errors** in /etc/php/7.2/fpm/php.ini <br/> <br/> (2) display_errors=on, **prioity is from highest to lowerest**: <br/> (2-1) **display_errors=on** in /etc/php/7.2/fpm/pool.d/www.conf <br/> (2-2) **display_errors=on** in /etc/php/7.2/apache2/php.ini <br/> <br/>Please notice this destination is not exclusive to a file path, you may have both.

Reviewer 2: ```suggestion
Client web page | **display_errors=on** in /etc/php/7.2/apache2/php.ini or in /etc/php/7.2/fpm/pool.d/www.conf | (1) **log_errors** in /etc/php/7.2/fpm/php.ini <br/> <br/> (2) display_errors=on, **priority is from highest to lowest**: <br/> (2-1) **display_errors=on** in /etc/php/7.2/fpm/pool.d/www.conf <br/> (2-2) **display_errors=on** in /etc/php/7.2/apache2/php.ini <br/> <br/>Please notice this destination is not exclusive to a file path, you may have both.
```

#### Loglevel supported
The mysqlnd_azure extension supports following log levels:
- [DEBUG]:
  1. source code path (inside which function);
  2. connection information (except password);
  3. verbose running status (e.g. customer given redirected server information directly... )
- [INFO]:
  1. connection fallback when mysqlnd\_azure.enableRedirection=preferred with details (cannot get redirection
      info from OK packet/ try connect using redirect information failed , ...);
  2. cache operations
- [ERROR]:
  1. any operation may cause customer see connection break (e.x. original conn/ redirected\_conn
    established failed, ... and connection fall back when mysqlnd\_azure.enableRedirection = ON)

- [SYSTM]:
Log info for mysqlnd_azure Log Module itself, e.g. invalid file length:
> ```
> 2020-03-27 09:09:05 [SYSTM] Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log
> ```

> note: out of memory errors are not covered by the log.

#### Level value relationship to level logged
- 0: No log written
- 1: [SYSTM] + [ERROR]
- 2: [SYSTM] + [ERROR] + [INFO]
- 3: [SYSTM] + [ERROR] + [INFO] + [DEBUG]

## Usage Example

> > You can add to section [mysqlnd\_azure] in file `php.ini` as follows:

Reviewer 2: Can we add some more details about the example?

```suggestion
> You can add to section [mysqlnd\_azure] in file `php.ini`. In the below example, the logs are output to "/tmp/test.log" (set by logfilePath and using logOutput = 2). The logging level includes system, error, info, and debug logs (set using logLevel = 3).
```
